File: Graph_Algorithms/Counting Rooms/sol.cpp
/***********************************************
 *                                             *
 *   author: raghav_0901 (Raghav Dalmia)       *
 *   updated on: 03-10-2020,  1:26:13 am IST   *
 *                                             *
\***********************************************/

#include <bits/stdc++.h>
using namespace std;

const int nx = (1e3) + 1;
int n, m;

bool vis[nx][nx];
vector<vector<int>> mov = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

bool valid(int r, int c) { return (r >= 0 && r < n && c >= 0 && c < m); }

void dfs(vector<string> &v, int r, int c)
{
    vis[r][c] = 1;
    for (auto it : mov)
    {
        int nr = r + it[0];
        int nc = c + it[1];
        if (valid(nr, nc) && !vis[nr][nc] && v[nr][nc] == '.')
            dfs(v, nr, nc);
    }
}

int main( )
{
    cin >> n >> m;
    vector<string> v;
    for (int i = 0; i < n; i++)
    {
        string s;
        cin >> s;
        v.emplace_back(s);
        for (int j = 0; j < m; j++) vis[i][j] = 0;
    }
    int ans = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            if (!vis[i][j] && v[i][j] == '.')
            {
                ans++;
                dfs(v, i, j);
            }
    cout << ans;
    return 0;
}

File: Graph_Algorithms/Counting Rooms/sol.py
################################################
#                                              #
#   author: raghav_0901 (Raghav Dalmia)        #
#   updated on: 03-10-2020,  1:40:05 am IST    #
#                                              #
################################################
 
from math import *
 
 
def _in():
    return [int(x) for x in input().split(' ')]
 
 
mov = ((1, 0), (-1, 0), (0, 1), (0, -1))
 
 
def valid(r, c):
    return (r >= 0 and r < n and c >= 0 and c < m)
 
 
def dfs(r, c):
    vis[r][c] = True
    for (x, y) in mov:
        nr = r + x
        nc = c + y
        if(valid(nr, nc) and vis[nr][nc] == False and arr[nr][nc] == '.'):
            dfs(nr, nc)
 
 
if __name__ == "__main__":
    global n, m, arr, vis
    n, m = _in()
    arr, vis = [], []
    for _ in range(n):
        s = input()
        arr.append(s)
        vis.append([False] * m)
    ans = 0
    for i in range(n):
        for j in range(m):
            if(vis[i][j] == False and arr[i][j] == '.'):
                ans = ans + 1
                dfs(i, j)
    print(ans)
